src/Npgsql/TypeHandlers/UnmappedCompositeHandler.cs

```diff
@@ -192,10 +192,13 @@ protected internal override Task WriteObjectWithLength(object value, NpgsqlWrite
                 ? WriteWithLengthInternal<DBNull>(null, buf, lengthCache, parameter, async)
                 : WriteWithLength(value, buf, lengthCache, parameter, async);
 
-        protected override Task WriteWithLength<T2>(T2 value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
+        protected override async Task WriteWithLength<T2>(T2 value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
         {
+            if (buf.WriteSpaceLeft < 4)
```

Comment:
Please take note of this comment on [`NpgsqlTypeHandler<T>.WriteWithLength()`](https://github.com/npgsql/Npgsql/blob/df65e0a42d2f5efa4f15df7d5da0c0dd07985a46/src/Npgsql/TypeHandling/NpgsqlTypeHandler%60.cs#L179):

> Note that this method assumes it can write 4 bytes of length (already verified by WriteWithLengthInternal<TAny>)

As long as any write goes through `WriteWithLengthInternal()`, we should be fine... @YohDeadfall can you please describe exactly which scenario triggers the error?

To be honest, looking at this split design (`WriteWithLengthInternal`/`WriteWithLength`) I'm not sure what the logic here was. Maybe it was simply to avoid duplicating the 3 lines of `WriteWithLength` twice inside `WriteWithLengthInternal` (once in the async/long local function, once in the non-async optimized version). If so it doesn't make much sense - we can fold them into one.

There's also a naming problem: the "long" local function inside `WriteWithLengthInternal` is called `WriteWithLengthLong`, the `Internal` is missing. But I'd like to understand @YohDeadfall's real issue before refactoring anything...

Reply:
> @YohDeadfall can you please describe exactly which scenario triggers the error?

In my case an array of composites sometimes triggers the error. It happened not always and depending on the data.

> If so it doesn't make much sense - we can fold them into one.

I've already removed code duplication in the [YohDeadfall:handler-devirtualization](https://github.com/YohDeadfall/npgsql/tree/handler-devirtualization) branch. But the change doesn't fit the hotfix, it's too large.

Reply:
I still don't understand... Can you please post what you think is causing this bug in the current version of Npgsql? As I wrote before, writes *should* be going through `WriteWithLengthInternal()`, which already does this check. So how can this be happening exactly?

Reply:
Everything seems to be clear here, I'll mark this conversation resolved (and am looking forward to seeing your refactoring!).